site/app/libraries/NotebookUtils.php

```diff
@@ -5,90 +5,254 @@
+            $warning_cell = [
+                'type' => 'markdown',
+                'markdown_data' => '### Notebook Rendered with Skipped Content' . PHP_EOL .
+                'Some content was skipped due to size limits. ' .
+                ($skipped_content > 0 ? "Skipped $skipped_content attachment(s)." : '') .
+                ($skipped_output > 0 ? " Skipped $skipped_output output(s)." : '') .
+                PHP_EOL . 'Download the notebook to view the full notebook.'
+            ];
+            array_unshift($cells, $warning_cell);
+        }
+
+        return $cells;
+    }
+
+    /**
+     * Process a markdown cell and return the processed cell.
+     *
+     * @param array<string,mixed> $cell
+     * @return array{
+     *     cell: array<string,mixed>,
+     *     skipped_content: int
+     * }
+     */
+    private static function processMarkdownCell(array $cell): array {
+        $skipped_content = 0;
+        $markdown = is_array($cell['source']) ? implode($cell['source']) : (string) $cell['source'];
+        $search = [];
+        $replace = [];
+
+        // Render attachment images (not HTML)
+        if (isset($cell['attachments']) && count($cell['attachments']) > 0) {
+            // Add attachments and a corresponding string to replace the attachment
+            foreach ($cell['attachments'] as $filename => $attachment) {
+                foreach ($attachment as $mime => $base64) {
+                    $search[] = "attachment:$filename";
+                    if (!in_array($mime, self::MIME_TYPES, true)) {
+                        $replace[] = 'Image skipped: image type not supported.';
+                        $skipped_content += 1;
+                    }
+                    elseif (strlen($base64) > self::IMG_SIZE_LIMIT) {
+                        $replace[] = 'Image skipped: exceeds size limit of ' . (self::IMG_SIZE_LIMIT / (1024 * 1024)) . ' MB. Download the notebook to view the image.';
+                        $skipped_content += 1;
+                    }
+                    else {
+                        $data_uri = 'data:' . $mime . ";base64," . $base64;
+                        $replace[] = $data_uri;
+                    }
+                }
+            }
+        }
+        // Replace all attachments with the data URI or a message if skipped
+        $markdown = str_replace($search, $replace, $markdown);
+        return [
+            'cell' => [
+                'type' => 'markdown',
+                'markdown_data' => $markdown,
+            ],
+            'skipped_content' => $skipped_content
+        ];
+    }
+
+    /**
+     * Process a code cell and return the processed cell.
+     *
+     * @param array<string,mixed> $cell
+     * @param array<string,mixed> $filedata
+     * @return array{
+     *     cells: array<int,array<string,mixed>>,
+     *     skipped_output: int
+     * }
+     */
+    private static function processCodeCell(array $cell, array $filedata): array {
```

Comment:
What exactly does it mean to "process" a code cell?  Perhaps rename this function?